### PR TITLE
Refactor UniqueEntityList 

### PR DESCRIPTION
**Pull request by ZhangAnli**

### CODE REVIEW

**Comment by ZhangAnli** [comment]
Refactor UniqueEntityList to remove unnecessary methods that use Entity parameters.

File: src/test/java/dog/pawbook/model/managedentity/owner/UniqueEntityListTest.java [hunk 1]
```diff
@@ -70,20 +71,20 @@ public void setOwner_targetOwnerNotInList_throwsOwnerNotFoundException() {
...
-        assertThrows(EntityNotFoundException.class, () -> uniqueEntityList.remove(ALICE));
+        assertThrows(EntityNotFoundException.class, () -> uniqueEntityList.remove(-1));
     }
 
     @Test
     public void remove_existingOwner_removesOwner() {
         uniqueEntityList.add(ALICE);
-        uniqueEntityList.remove(ALICE);
+
+        List<Pair<Integer, Entity>> targets = uniqueEntityList.asObservableList().stream()
```
Comment (kouyk) [inline src/test/java/dog/pawbook/model/managedentity/owner/UniqueEntityListTest.java:83]: okay I think i get the rationale for the returning the internalList directly, but it is alright to use the unmodifiable version as many other tests does the same. Also I think if this is targeted at owners but will not test dogs, it is a good idea to filter using instanceof as well.

File: src/test/java/dog/pawbook/storage/JsonAddressBookStorageTest.java [hunk 2]
```diff
@@ -73,7 +77,11 @@ public void readAndSaveAddressBook_allInOrder_success() throws Exception {
 
         // Modify data, overwrite exiting file, and read back
         original.addEntity(HOON);
-        original.removeEntity(ALICE);
+        List<Pair<Integer, Entity>> targets = original.getEntityList().stream()
+                .filter(p -> p.getValue().isSameEntity(ALICE))
+                .collect(toList());
+        int aliceId = targets.get(0).getKey();
+        original.removeEntity(aliceId);
```
Comment (kouyk) [inline src/test/java/dog/pawbook/storage/JsonAddressBookStorageTest.java:84]: same the other one

File: src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java [hunk 3]
```diff
@@ -189,6 +173,13 @@ public void setEntities(List<Pair<Integer, Entity>> entities) {
         return internalUnmodifiableList;
     }
 
+    /**
+     * Returns the backing list as an {@code ObservableList}.
+     */
+    public ObservableList<Pair<Integer, Entity>> asObservableList() {
+        return internalList;
+    }
+
```
[the remark this answers is not in the document]
Comment (kouyk) [reply src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java:182]: Okay I understand this portion now, it might seem fine to do this for now, but the issue is that this codebase isn't very small, having equivalent api will confuse others that are not so familiar with this portion, and any direct modification is now possible since Java passes by reference, it will be difficult to track down bugs if someone modifies it and violates the assumptions we have about unique entities etc.

File: src/test/java/dog/pawbook/model/managedentity/owner/UniqueEntityListTest.java [hunk 4]
```diff
@@ -70,20 +71,20 @@ public void setOwner_targetOwnerNotInList_throwsOwnerNotFoundException() {
...
 
-    @Test
-    public void remove_nullOwner_throwsNullPointerException() {
-        assertThrows(NullPointerException.class, () -> uniqueEntityList.remove(null));
-    }
-
     @Test
     public void remove_ownerDoesNotExist_throwsOwnerNotFoundException() {
-        assertThrows(EntityNotFoundException.class, () -> uniqueEntityList.remove(ALICE));
+        assertThrows(EntityNotFoundException.class, () -> uniqueEntityList.remove(-1));
```
[the remark this answers is not in the document]
Comment (kouyk) [reply src/test/java/dog/pawbook/model/managedentity/owner/UniqueEntityListTest.java:76]: this can be another test to remove an invalid ID equivalent to the null test above.

**Review by kouyk — CHANGES_REQUESTED** [review]
Can I trouble you to update all the Javadocs to make them consistent by referring to Entity as much as possible? There are several changes we need to discuss as well.

File: src/main/java/dog/pawbook/model/AddressBook.java [hunk 5]
```diff
@@ -57,17 +57,21 @@ public void resetData(ReadOnlyAddressBook newData) {
         setEntities(newData.getEntityList());
     }
 
-    //// owner-level operations
+    // owner-level operations
```
[the remark this answers is not in the document]
Comment (ZhangAnli) [reply src/main/java/dog/pawbook/model/AddressBook.java:60]: Edited

File: src/main/java/dog/pawbook/model/AddressBook.java [hunk 6]
```diff
@@ -57,17 +57,21 @@ public void resetData(ReadOnlyAddressBook newData) {
...
     public boolean hasEntity(Entity entity) {
         requireNonNull(entity);
         return entities.contains(entity);
     }
 
+    /**
+     * Returns true if a owner with the same identity as {@code owner} exists in the address book.
+     */
     public boolean hasEntity(int id) {
+        requireNonNull(id);
```
[the remark this answers is not in the document]
Comment (ZhangAnli) [reply src/main/java/dog/pawbook/model/AddressBook.java:74]: Edited

File: src/main/java/dog/pawbook/model/AddressBook.java [hunk 7]
```diff
@@ -94,13 +98,13 @@ public void setEntity(int targetId, Entity editedEntity) {
...
-    public void removeEntity(Entity key) {
-        entities.remove(key);
-    }
+    //    /**
+    //     * Removes {@code key} from this {@code AddressBook}.
+    //     * {@code key} must exist in the address book.
+    //     */
+    //    public void removeEntity(Entity key) {
+    //      entities.remove(key);
+    //    }
```
[the remark this answers is not in the document]
Comment (ZhangAnli) [reply src/main/java/dog/pawbook/model/AddressBook.java:107]: Edited

File: src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java [hunk 8]
```diff
@@ -34,22 +34,31 @@
...
+        requireNonNull(entity);
+        return internalList.stream().anyMatch(p -> p.getValue().isSameEntity(entity));
+    }
+
     /**
      * Retrieve the index of an entity stored in the internal list.
      * @return the index if found, -1 otherwise.
      */
-    private int getIndexOf(int id) {
+    public int getIndexOf(int id) {
```
[the remark this answers is not in the document]
Comment (ZhangAnli) [reply src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java:61]: Edited

File: src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java [hunk 9]
```diff
@@ -137,25 +135,11 @@ public void setEntity(int targetId, Entity editedEntity) {
...
-        }
-
-        internalList.remove(index);
-    }
-
     /**
      * Removes the entity with the given ID.
      */
     public void remove(int toRemoveId) {
+        requireNonNull(toRemoveId);
```
[the remark this answers is not in the document]
Comment (ZhangAnli) [reply src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java:142]: Edited

File: src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java [hunk 10]
```diff
@@ -93,7 +87,10 @@ private int genID() {
      */
     public void add(Entity toAdd) {
         requireNonNull(toAdd);
-        if (contains(toAdd)) {
+
+        boolean containsToAdd = internalList.stream().anyMatch(p -> p.getValue().equals(toAdd));
+
+        if (containsToAdd) {
```
[the remark this answers is not in the document]
Comment (ZhangAnli) [reply src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java:93]: Edited

File: src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java [hunk 11]
```diff
@@ -105,7 +102,7 @@ public void add(Entity toAdd) {
      */
     public void add(Entity toAdd, int id) {
         requireNonNull(toAdd);
-        if (contains(toAdd) || contains(id)) {
+        if (contains(id)) {
```
[the remark this answers is not in the document]
Comment (ZhangAnli) [reply src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java:105]: Edited

File: src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java [hunk 12]
```diff
@@ -34,22 +34,31 @@
...
+     * Checks if list contains a particular owner
+     *
+     * @param entity owner entity
+     * @return boolean of whether owner exists.
+     */
+    public boolean contains(Entity entity) {
+        requireNonNull(entity);
+        return internalList.stream().anyMatch(p -> p.getValue().isSameEntity(entity));
+    }
+
```
[the remark this answers is not in the document]
Comment (ZhangAnli) [reply src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java:56]: Edited

File: src/test/java/dog/pawbook/model/managedentity/owner/UniqueEntityListTest.java (hunk 1 above)
Comment (ZhangAnli) [reply src/test/java/dog/pawbook/model/managedentity/owner/UniqueEntityListTest.java:83]: Edited

File: src/test/java/dog/pawbook/storage/JsonAddressBookStorageTest.java (hunk 2 above)
Comment (ZhangAnli) [reply src/test/java/dog/pawbook/storage/JsonAddressBookStorageTest.java:84]: I think for this I didn't use the asObservableList method

File: src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZhangAnli) [reply src/main/java/dog/pawbook/model/managedentity/owner/UniqueEntityList.java:182]: Removed

File: src/test/java/dog/pawbook/model/managedentity/owner/UniqueEntityListTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ZhangAnli) [reply src/test/java/dog/pawbook/model/managedentity/owner/UniqueEntityListTest.java:76]: Edited

**Review by kouyk — APPROVED** [review]
Looks good, just that John's current PR will be rather useful in that addEntity or any real add method will return the ID of the new entity, that will help to simply the test methods greatly. You can create an issue for this and also that now we use ID in most places, we can remove the use of isSameEntity and just use the stronger equals.
